File: src/main/java/com/sedmelluq/discord/lavaplayer/natives/opus/OpusDecoderLibrary.java
package com.sedmelluq.discord.lavaplayer.natives.opus;

import com.sedmelluq.discord.lavaplayer.natives.ConnectorNativeLibLoader;

import java.nio.ByteBuffer;
import java.nio.ShortBuffer;

class OpusDecoderLibrary {
  private OpusDecoderLibrary() {

  }

  static OpusDecoderLibrary getInstance() {
    ConnectorNativeLibLoader.loadConnectorLibrary();
    return new OpusDecoderLibrary();
  }

  native long create(int sampleRate, int channels);

  native void destroy(long instance);

  native int decode(long instance, ByteBuffer directInput, int inputSize, ShortBuffer directOutput, int frameSize);
}

File: main/src/main/java/com/sedmelluq/discord/lavaplayer/natives/opus/OpusDecoderLibrary.kt
package com.sedmelluq.discord.lavaplayer.natives.opus

import com.sedmelluq.discord.lavaplayer.natives.ConnectorNativeLibLoader.loadConnectorLibrary
import java.nio.ByteBuffer
import java.nio.ShortBuffer

internal class OpusDecoderLibrary private constructor() {
    external fun create(sampleRate: Int, channels: Int): Long
    external fun destroy(instance: Long)
    external fun decode(
        instance: Long,
        directInput: ByteBuffer?,
        inputSize: Int,
        directOutput: ShortBuffer?,
        frameSize: Int
    ): Int

    companion object {
        @JvmStatic
        val instance: OpusDecoderLibrary
            get() {
                loadConnectorLibrary()
                return OpusDecoderLibrary()
            }
    }
}
